File: examples/detectingHPCodePattern.cpp
#include "Vernier.hpp"

using namespace vernier;
using namespace cv;
using namespace std;

/** This example shows how to estimate the pose of HP code markers in an image.
 * 
 * The method is described in the following paper: 
 * 
 * [1] A. N. Andre et al., “Automating Robotic Micro-Assembly of Fluidic Chips 
 * and Single Fiber Compression Tests Based-on ΘVisual Measurement With 
 * High-Precision Fiducial Markers, IEEE Transactions on Automation Science and 
 * Engineering, 2022.
 */
int main() {

    // Loading the image
    string filename = "HPCodeImage_20um.png";
    Mat image = cv::imread(filename);

    // Detecting the pattern and estimating its pose
    double physicalPeriod = 20; // µm
    int snapshotSize = 256; // pixels
    HPCodePatternDetector detector(physicalPeriod, 33, snapshotSize);
    detector.compute(image);

    // Displaying the pose if a pattern has been found
    cout << "Detector: " << detector.toString() << endl;
    if (detector.patternFound()) {
        cout << "Found " << detector.markers.size() << " markers in " << filename << endl;
        for (auto it : detector.markers) {
            cout << "  Stamp " << it.first << " at " << it.second.toString() << endl;
        }
        detector.draw(image);
    } else {
        cout << "Marker not found..." << endl;
    }

    // Showing image and its spectrum
    detector.showControlImages();
    imshow(filename, image);
    waitKey(3000);
}

File: python/examples/detectingHPCodePattern.py
#
# This file is part of the VERNIER Library.
#
# Copyright (c) 2018-2025 CNRS, ENSMM, UMLP.
#

"""Python port of examples/detectingHPCodePattern.cpp.

This example shows how to estimate the pose of HP code markers in an image.

The method is described in the following paper:

[1] A. N. Andre et al., "Automating Robotic Micro-Assembly of Fluidic Chips
and Single Fiber Compression Tests Based-on Visual Measurement With
High-Precision Fiducial Markers", IEEE Transactions on Automation Science and
Engineering, 2022.

Run it from the build directory, where the module and the images live:

    cd build/python && python3 detectingHPCodePattern.py [--show]
"""

import sys

import pyvernier as vernier

# Loading the image
filename = "HPCodeImage_20um.png"
image = vernier.readImage(filename)

# Detecting the pattern and estimating its pose
physicalPeriod = 20  # µm
snapshotSize = 256  # pixels
detector = vernier.HPCodePatternDetector(physicalPeriod, 33, snapshotSize)
detector.compute(image)

# Displaying the pose if a pattern has been found
print("Detector:", detector)
if detector.patternFound():
    # `markers` is a dict {code id: Pose}.
    print("Found %d markers in %s" % (len(detector.markers), filename))
    for code_id, pose in sorted(detector.markers.items()):
        print("  Stamp %d at %s" % (code_id, pose))
    image = detector.draw(image)
else:
    print("Marker not found...")

# Showing image and its spectrum (needs a display, hence the flag)
if "--show" in sys.argv:
    detector.showControlImages()
    vernier.showImage(filename, image)
    vernier.waitKey(3000)
